From da7ffbd0c030751469ffdcb5ac99b61f0452b850 Mon Sep 17 00:00:00 2001
From: Konrad Leszczynski <konrad.leszczynski@intel.com>
Date: Tue, 15 Sep 2015 16:11:00 +0100
Subject: [PATCH] Send zero packet for intel 7260 modem

For Intel 7260 modem, it is needed for host side to send zero packet
if the BULK OUT size is equal to USB endpoint max packet length.
Otherwise, modem side may still wait for more data and cannot give
response to host side.

Tracked-On: https://jira01.devtools.intel.com/browse/IMINAN-11648
Tracked-On: https://jira01.devtools.intel.com/browse/GMINL-17076
Change-Id: Id213bf0a425e3a6b511f500f05b9cdb37e4e9323
Signed-off-by: Konrad Leszczynski <konrad.leszczynski@intel.com>
Reviewed-on: https://android.intel.com:443/304374
---
 drivers/usb/class/cdc-acm.c | 11 ++++++++++-
 drivers/usb/class/cdc-acm.h |  7 +++++++
 2 files changed, 17 insertions(+), 1 deletion(-)

diff --git a/drivers/usb/class/cdc-acm.c b/drivers/usb/class/cdc-acm.c
index 3c9b132..e37d4f1 100644
--- a/drivers/usb/class/cdc-acm.c
+++ b/drivers/usb/class/cdc-acm.c
@@ -885,7 +885,7 @@ static int wait_serial_change(struct acm *acm, unsigned long arg)
 		}
 	} while (!rv);
 
-	
+
 
 	return rv;
 }
@@ -1046,6 +1046,13 @@ static int acm_write_buffers_alloc(struct acm *acm)
 	return 0;
 }
 
+/* Zero packet needs to be sent for some modems */
+static bool acm_needs_zerp_pkt(struct usb_device *usb_dev)
+{
+	return usb_dev->descriptor.idVendor == INTEL_MODEM_VID &&
+		usb_dev->descriptor.idProduct == INTEL_MODEM_PID;
+}
+
 static int acm_probe(struct usb_interface *intf,
 		     const struct usb_device_id *id)
 {
@@ -1384,6 +1391,8 @@ made_compressed_probe:
 				usb_sndbulkpipe(usb_dev, epwrite->bEndpointAddress),
 				NULL, acm->writesize, acm_write_bulk, snd);
 		snd->urb->transfer_flags |= URB_NO_TRANSFER_DMA_MAP;
+		if (acm_needs_zerp_pkt(usb_dev))
+			snd->urb->transfer_flags |= URB_ZERO_PACKET;
 		snd->instance = acm;
 	}
 
diff --git a/drivers/usb/class/cdc-acm.h b/drivers/usb/class/cdc-acm.h
index 80826f8..07780e9 100644
--- a/drivers/usb/class/cdc-acm.h
+++ b/drivers/usb/class/cdc-acm.h
@@ -48,6 +48,13 @@
 #define ACM_CTRL_OVERRUN	0x40
 
 /*
+ * Intel 7260 modem VID/PID.
+ */
+
+#define INTEL_MODEM_VID		0x1519
+#define INTEL_MODEM_PID		0x0452
+
+/*
  * Internal driver structures.
  */
 
-- 
1.9.1

